### DISCUSSION TITLE
Batch loading filtered/paginated relationships

### DISCUSSION THREAD

**Discussion by MeRuslan** [sqlalchemy/sqlalchemy#7678] [Usage Questions] [answered]
I'm currently developing graphql-sqlalchemy integration, and am considering approaches to batch loading relationships, with filtering and pagination applied.
Loading them without filtering is doable and I've found a way, but with filtering, it becomes quite a task, cause I can't slice and dice selectinload, I'd have to do that manually.
Currently I'm leaning towards using 'lateral', which seems to be the only way to do that in SQL.

A sample to illustrate the point: 
```
subq = select(Messages).filter(Messages.user_id == User.id).order_by(Messages.date.desc()).limit(10).subquery().lateral()
q = select(User).outerjoin(subq).options(contains_eager(User.messages, alias=subq))

```
Here `.filter(Messages.user_id == User.id)` is basically applying a primary join condition on the user-message relationship.
I need this filter to work for a generic case, for any kind of a relationship.

Thus, the question is, is there an idiomatic way to apply filtering on children for any given relationship. Including many-to-many.

**Comment by zzzeek** [comment]
hey there -

yes there is, after a long time of this not being a thing, people really want to have relationships that are loaded on the parent object with filtering applied, so take a look at the feature described at https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#adding-criteria-to-loader-options where you can add that criteria in-place to any loader option, as well as https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.with_loader_criteria which will apply a criteria throughout a query.

**Reply by MeRuslan** [reply]
For anyone who might visit, please see the discussion below as well.

**Comment by MeRuslan** [comment]
@zzzeek thank you very much!

It does solve the filtering part, but not paginating. I'm not sure loader options allow pagination.

I'm afraid that for pagination on children to be applied, I'd have to use lateral, and a separate subquery for children (please let me know if I'm wrong here). 
If that's the case, then I'll need to filter children as to join them to the parent objects I need prior to filtering, paginating children.

**Reply by zzzeek** [reply]
there's many ways to do pagination.  I'm not sure where the "lateral" part comes in.    We have a pretty decent "pagination" recipe based on window functions which you can see here: https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#row-limited-relationships-with-window-functions which should be able to be adapted to the "and_()" feature mentioned above.

**Reply by MeRuslan** [reply]
Yes! Window functions were the other one I was considering, and I completely missed that I can use aforementioned "and_" feature, thank you very much :D

**Reply by zzzeek** [reply]
great, good luck getting it going that's pretty advanced.    There were more options here too, since you can also manually "attach" related collections to objects too if you wanted to go that route.

**Reply by MeRuslan** [reply]
Although, it is not as generic of a solution as the "lateral" one, since here I need to predefine ordering and a limit at the relationship level.
I wanted organize it in a way, that the user can define these.

Maybe this approach will be flexible enough, I might need to try it out. As not that often do you need the fine-controlled slicing of children. Maybe we could get away with predefined ones, cause generating relationships to fit requests seems very wrong.

**Reply by MeRuslan** [reply]
> great, good luck getting it going that's pretty advanced. There were more options here too, since you can also manually "attach" related collections to objects too if you wanted to go that route.

Yeah, exactly.
And thanks!

**Comment by MeRuslan** [comment, marked as answer]
I've ended up using lateral after all, had to come back to it. On the plus side, now I know how to marry lateral with complex queries and ORM.
Didn't like paginating with partitioning over stuff, since it's hard defined, thus troublesome, even more so when trying different pagination/filtering options.

Here's sample code
```python 
# form lateral query
query = select(related_model).filter(relationship_property.primaryjoin)
# filter / order / limit as you wish
query = query.limit(10).offset(0)
query = query.lateral()
# alias the related model from lateral query to get it, and not the parent one
ra = aliased(related_model, query)
# filter the parent to only the ones we need, but return aliased related model
result_q = (
    select(ra).select_from(parent_model)
              .outerjoin(query, true())
              .where(parent_model.id.in_(ids))
)
```
I've tested out a few generated queries similar to the one presented, postgres seems to "prefer" lateral ones as well.
